### ISSUE TITLE
Kubuntu large icon task switcher

### ISSUE THREAD

**Issue by bibendi** [telegramdesktop/tdesktop#7917]
### Steps to reproduce
1. Go to System Settings - Window Managment - Task Switcher
2. Choose Large Icons
3. Change window by pressing Alt+Tab

### Expected behaviour

Telegram icon is not blurred

### Actual behaviour

Telegram icon is blurred

### Configuration

**Operating system:** Kubuntu 20.04

**Version of Telegram Desktop:** 2.1.5

**Installation source (Linux Only)** - the official website / GitHub releases / flatpak / snap / distribution package: Website

![Screenshot_20200520_113259](https://user-images.githubusercontent.com/320886/82424428-d7371900-9a8d-11ea-9ed6-7e39a6845a28.png)

**Comment by ilya-fedin** [comment]
Looks like your icon theme has icons for all the apps on screenshot except Telegram

**Comment by bibendi** [comment]
I think this link would be very helpful https://bugs.kde.org/show_bug.cgi?id=351055

It can be fixed by the application itself.

**Comment by ilya-fedin** [comment]
> It can be fixed by the application itself.

Please, TL;DR, what is needed?

**Comment by bibendi** [comment]
An app should provide correct/good icon to kwin.

Please, read comments 34-35
https://bugs.kde.org/show_bug.cgi?id=351055#c34

**Comment by ilya-fedin** [comment]
> An app should provide correct/good icon to kwin.

tdesktop provides 256px icon. Is this good/correct?

**Comment by bibendi** [comment]
Unfortunately, I don't know :(

**Comment by bibendi** [comment]
I've found more information about the bug.

https://www.reddit.com/r/kde/comments/9kz6gv/please_make_kwin_automatically_recognise_correct/

TL;DR:

- An application should conform to the XDG icon theme standard. 

https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html

- A user can fix it for himself: 
1) Right click on title bar > more actions > special application settings > Appearances and fixes
2) Enable Desktop file name + Select "Force" and specify the desktop file name

**Comment by ilya-fedin** [comment]
> An application should conform to the XDG icon theme standard.

tdesktop installs icon to ~/.local/share/icons

**Comment by bibendi** [comment]
Okay, it seems, I've found the right answer.
The window should set `_NET_WM_ICON` property.
https://specifications.freedesktop.org/wm-spec/1.3/ar01s05.html

`_NET_WM_ICON CARDINAL[][2+n]/32`

> This is an array of possible icons for the client. This specification does not stipulate what size these icons should be, but individual desktop environments or toolkits may do so. The Window Manager MAY scale any of these icons to an appropriate size.

> This is an array of 32bit packed CARDINAL ARGB with high byte being A, low byte being B. The first two cardinals are width, height. Data is in rows, left to right and top to bottom. 

You can inspect the properties of other application windows by using the `xprop` CLI.

**Comment by bibendi** [comment]
Also, I have a problem with the tray icon.

![image](https://user-images.githubusercontent.com/320886/83126206-2d850700-a0e1-11ea-8d25-9f51bc998f06.png)

But it's a cool story of another investigation :)
Maybe you know where that icon is located to save my time?

**Comment by ilya-fedin** [comment]
> The window should set `_NET_WM_ICON` property.

It is set, otherwise there wouldn't be any icon

**Comment by ilya-fedin** [comment]
> Also, I have a problem with the tray icon.

#7399

**Comment by bibendi** [comment]
> It is set, otherwise there wouldn't be any icon

Yes, but there is only one 48x48 icon. Firefox has six icons.

**Comment by ilya-fedin** [comment]
> Yes, but there is only one 48x48 icon. Firefox has six icons.

As you can see, 256px icon is set. If Qt resizes it, tdesktop can't do anything with that.
https://github.com/ilya-fedin/tdesktop/blob/set-parent-only-on-wayland/Telegram/SourceFiles/window/main_window.cpp#L301
https://github.com/ilya-fedin/tdesktop/blob/set-parent-only-on-wayland/Telegram/SourceFiles/window/main_window.cpp#L109
https://github.com/ilya-fedin/tdesktop/blob/set-parent-only-on-wayland/Telegram/SourceFiles/core/application.h#L119
https://github.com/ilya-fedin/tdesktop/blob/set-parent-only-on-wayland/Telegram/SourceFiles/core/application.cpp#L100
https://github.com/ilya-fedin/tdesktop/blob/set-parent-only-on-wayland/Telegram/SourceFiles/window/main_window.cpp#L52

**Comment by bibendi** [comment]
I googled more and what I found out.
It's a specific problem of integration Qt applications with KDE.
Here the same issue https://github.com/CDrummond/cantata/issues/1206
And here the workaround https://github.com/CDrummond/cantata/commit/73b48c34894f486241a3dbdba14a7bbfca130a10

**Comment by ilya-fedin** [comment]
But he just added a 128px, tdesktop already sets 256px png.

**Comment by bibendi** [comment]
Ok. I give up. I hope there would be someone more opinionated in Qt than me, who digs deeper into Cantana's master branch and find the workaround. Let's leave this issue open.

**Comment by ilya-fedin** [comment]
I did some testing. 128x128px icon appears in xprop only with KDE platfromtheme plugin, that can't be used with binary from github. Sorry.

**Comment by ilya-fedin** [comment]
@Aokromes can you close this as 3rd party bug?

**Comment by ilya-fedin** [comment]
Looks like I found where Qt bug is placed
https://github.com/qt/qtbase/blob/abfb1b8/src/platformsupport/themes/genericunix/qgenericunixthemes.cpp#L248
https://github.com/KDE/plasma-integration/blob/4a7786999d318759628b9cc3ea1ff963547b561f/src/platformtheme/khintssettings.cpp#L147

**Comment by ilya-fedin** [comment]
Well, I tried to patch Qt and icon is 48x48 anyway

**Comment by ilya-fedin** [comment]
Icon is OK on snap
Icon is OK on flatpak
And only static binary is blurred

**Comment by ilya-fedin** [comment]
Well, I found the root of the problem. It is /usr/share/icons/breeze/apps/48/telegram.svg. KDE developers for some reason provide telegram icon in breeze only with 48px size that overrides telegram's 256px icon. Probably, you should open a bug in KDE bugtracker.

**Comment by bibendi** [comment]
Thank you so much!

https://bugs.kde.org/show_bug.cgi?id=422253
